File: app/src/main/java/com/qurantv/app/domain/IslamicHiliteRects.kt
package com.qurantv.app.domain

/**
 * Computes exact per-ayah highlight rects for islamic.app mushaf pages using
 * the same region logic as the KSU site's hilitePage(): an ayah's highlight
 * spans from where the previous ayah ENDED to where this ayah ends, drawn as
 * up to three rects (its first partial line from the line's left edge to the
 * previous end, its last partial line from its end to the line's right edge,
 * and full-width middle lines). Same-line ayahs collapse to a single rect.
 *
 * The ayah-end marker (۝ + digits) is the LAST tspan content of each ayah, so
 * the highlight ends exactly at the ayah number and never bleeds into the next
 * ayah's text on a shared line.
 *
 * Text width measurement is injected ([measure]) so the algorithm stays pure
 * (unit-testable); the loader supplies Android `Paint.measureText`.
 */
object IslamicHiliteRects {

    data class LineLayout(
        val lineTop: Float,
        val lineBottom: Float,
        val lineLeft: Float,
        val lineRight: Float,
    )

    /** ayah key ("s:a") -> end position (x, baseline) in document order. */
    private data class AyahEnd(val key: String, val x: Float, val y: Float)

    /**
     * @param lines the page's ayah text lines (from [IslamicPageBands.parseLines])
     * @param viewBoxWidth /[viewBoxHeight] the SVG viewBox (fraction space)
     * @param measure text -> width for the line's font (Android Paint.measureText)
     * @return ayah key -> highlight rects in fraction-of-page units
     */
    fun build(
        lines: List<IslamicLine>,
        viewBoxWidth: Float,
        viewBoxHeight: Float,
        measure: (text: String, fontSize: Float) -> Float,
    ): Map<String, List<KsuHiliteGeometry.Rect>> {
        if (lines.isEmpty() || viewBoxWidth <= 0f || viewBoxHeight <= 0f) return emptyMap()

        // 1. Layout every line and record each ayah's end position.
        val ends = LinkedHashMap<String, AyahEnd>()
        val lineLayouts = ArrayList<LineLayout>(lines.size)
        for (line in lines) {
            val widths = line.tspans.map { measure(it.text, line.fontSize).coerceAtLeast(0f) }
            val total = widths.sum()
            val lineRight = line.anchorX + total / 2f
            val lineLeft = line.anchorX - total / 2f
            lineLayouts += LineLayout(
                lineTop = line.baselineY - line.fontSize * 0.95f,
                lineBottom = line.baselineY + line.fontSize * 0.35f,
                lineLeft = lineLeft,
                lineRight = lineRight,
            )
            // tspan j occupies [lineRight - sum_{i<=j} w, lineRight - sum_{i<j} w)
            var consumed = 0f
            for (j in line.tspans.indices) {
                consumed += widths[j]
                val key = line.tspans[j].ayahKey
                if (key != null) {
                    ends[key] = AyahEnd(key, lineRight - consumed, line.baselineY)
                }
            }
        }

        // 2. Build each ayah's region (from where it starts to its own end).
        val ayahKeys = ends.keys.toList()
        // end positions of every tspan by line, to know shared lines.
        val out = LinkedHashMap<String, List<KsuHiliteGeometry.Rect>>(ayahKeys.size)
        for (idx in ayahKeys.indices) {
            val key = ayahKeys[idx]
            val end = ends[key]!!
            val firstLineIndex = firstLineOf(lines, key)
            val lastLineIndex = lineIndexOf(lines, end.y)
            val firstLayout = lineLayouts[firstLineIndex]
            val lastLayout = lineLayouts[lastLineIndex]

            // Where the ayah begins on its first line: right after the previous
            // ayah when they share the line, else at the line's right edge.
            val prevEnd = if (idx == 0) null else ends[ayahKeys[idx - 1]]
            val startX = if (prevEnd != null && lineIndexOf(lines, prevEnd.y) == firstLineIndex) {
                prevEnd.x
            } else {
                firstLayout.lineRight
            }

            val rects = if (firstLineIndex == lastLineIndex) {
                // Single line: one rect between its start and end.
                listOf(
                    rect(startX, end.x, firstLayout.lineTop, firstLayout.lineBottom, viewBoxWidth, viewBoxHeight),
                )
            } else {
                val builder = ArrayList<KsuHiliteGeometry.Rect>()
                // First line: from the line's left edge to where the ayah begins.
                builder += rect(
                    firstLayout.lineLeft,
                    startX,
                    firstLayout.lineTop,
                    firstLayout.lineBottom,
                    viewBoxWidth,
                    viewBoxHeight,
                )
                // Middle lines: full width.
                for (m in firstLineIndex + 1 until lastLineIndex) {
                    val ml = lineLayouts[m]
                    builder += rect(ml.lineLeft, ml.lineRight, ml.lineTop, ml.lineBottom, viewBoxWidth, viewBoxHeight)
                }
                // Last line: from the ayah's end to the line's right edge.
                builder += rect(
                    end.x,
                    lastLayout.lineRight,
                    lastLayout.lineTop,
                    lastLayout.lineBottom,
                    viewBoxWidth,
                    viewBoxHeight,
                )
                builder
            }
            out[key] = rects
        }
        return out
    }

    /** The line holding this ayah's FIRST tspan. */
    private fun firstLineOf(lines: List<IslamicLine>, ayahKey: String): Int {
        for (i in lines.indices) {
            if (lines[i].tspans.any { it.ayahKey == ayahKey }) return i
        }
        return 0
    }

    private fun lineIndexOf(lines: List<IslamicLine>, baselineY: Float): Int {
        var best = 0
        var bestDist = Float.MAX_VALUE
        for (i in lines.indices) {
            val d = kotlin.math.abs(lines[i].baselineY - baselineY)
            if (d < bestDist) {
                bestDist = d
                best = i
            }
        }
        return best
    }

    private fun rect(
        left: Float,
        right: Float,
        top: Float,
        bottom: Float,
        vw: Float,
        vh: Float,
    ): KsuHiliteGeometry.Rect {
        val l = minOf(left, right)
        val r = maxOf(left, right)
        return KsuHiliteGeometry.Rect(
            left = (l / vw).coerceIn(0f, 1f),
            top = (top / vh).coerceIn(0f, 1f),
            right = (r / vw).coerceIn(0f, 1f),
            bottom = (bottom / vh).coerceIn(0f, 1f),
        )
    }
}

File: web/src/domain/IslamicHiliteRects.ts
// Ported 1:1 from app/src/main/java/com/qurantv/app/domain/IslamicHiliteRects.kt
// Exact per-ayah highlight rects for islamic.app pages (same region logic as
// the KSU hilitePage). Text width measurement is injected so the algorithm
// stays pure (the loader supplies a Paint.measureText equivalent).

import type { IslamicLine } from "./IslamicPageBands";
import type { KsuRect } from "./KsuHiliteGeometry";

interface LineLayout {
  lineTop: number;
  lineBottom: number;
  lineLeft: number;
  lineRight: number;
}

interface AyahEnd {
  key: string;
  x: number;
  y: number;
}

/**
 * @param lines the page's ayah text lines
 * @param viewBoxWidth / height the SVG viewBox (fraction space)
 * @param measure text -> width for the line's font
 * @returns ayah key -> highlight rects in fraction-of-page units
 */
export function buildIslamicRects(
  lines: IslamicLine[],
  viewBoxWidth: number,
  viewBoxHeight: number,
  measure: (text: string, fontSize: number) => number,
): Map<string, KsuRect[]> {
  if (lines.length === 0 || viewBoxWidth <= 0 || viewBoxHeight <= 0) return new Map();

  const ends = new Map<string, AyahEnd>();
  const lineLayouts: LineLayout[] = [];

  for (const line of lines) {
    const widths = line.tspans.map((t) => Math.max(0, measure(t.text, line.fontSize)));
    const total = widths.reduce((a, b) => a + b, 0);
    const lineRight = line.anchorX + total / 2;
    const lineLeft = line.anchorX - total / 2;
    lineLayouts.push({
      lineTop: line.baselineY - line.fontSize * 0.95,
      lineBottom: line.baselineY + line.fontSize * 0.35,
      lineLeft,
      lineRight,
    });
    let consumed = 0;
    for (let j = 0; j < line.tspans.length; j++) {
      consumed += widths[j];
      const key = line.tspans[j].ayahKey;
      if (key !== null) {
        ends.set(key, { key, x: lineRight - consumed, y: line.baselineY });
      }
    }
  }

  const ayahKeys = [...ends.keys()];
  const out = new Map<string, KsuRect[]>();
  for (let idx = 0; idx < ayahKeys.length; idx++) {
    const key = ayahKeys[idx];
    const end = ends.get(key)!;
    const firstLineIndex = firstLineOf(lines, key);
    const lastLineIndex = lineIndexOf(lines, end.y);
    const firstLayout = lineLayouts[firstLineIndex];
    const lastLayout = lineLayouts[lastLineIndex];

    const prevEnd = idx === 0 ? undefined : ends.get(ayahKeys[idx - 1]);
    const startX =
      prevEnd !== undefined && lineIndexOf(lines, prevEnd.y) === firstLineIndex
        ? prevEnd.x
        : firstLayout.lineRight;

    let rects: KsuRect[];
    if (firstLineIndex === lastLineIndex) {
      rects = [islamicRect(startX, end.x, firstLayout.lineTop, firstLayout.lineBottom, viewBoxWidth, viewBoxHeight)];
    } else {
      const builder: KsuRect[] = [];
      builder.push(
        islamicRect(firstLayout.lineLeft, startX, firstLayout.lineTop, firstLayout.lineBottom, viewBoxWidth, viewBoxHeight),
      );
      for (let m = firstLineIndex + 1; m < lastLineIndex; m++) {
        const ml = lineLayouts[m];
        builder.push(islamicRect(ml.lineLeft, ml.lineRight, ml.lineTop, ml.lineBottom, viewBoxWidth, viewBoxHeight));
      }
      builder.push(
        islamicRect(end.x, lastLayout.lineRight, lastLayout.lineTop, lastLayout.lineBottom, viewBoxWidth, viewBoxHeight),
      );
      rects = builder;
    }
    out.set(key, rects);
  }
  return out;
}

function firstLineOf(lines: IslamicLine[], ayahKey: string): number {
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].tspans.some((t) => t.ayahKey === ayahKey)) return i;
  }
  return 0;
}

function lineIndexOf(lines: IslamicLine[], baselineY: number): number {
  let best = 0;
  let bestDist = Number.MAX_VALUE;
  for (let i = 0; i < lines.length; i++) {
    const d = Math.abs(lines[i].baselineY - baselineY);
    if (d < bestDist) {
      bestDist = d;
      best = i;
    }
  }
  return best;
}

function islamicRect(left: number, right: number, top: number, bottom: number, vw: number, vh: number): KsuRect {
  const l = Math.min(left, right);
  const r = Math.max(left, right);
  return {
    left: clamp01(l / vw),
    top: clamp01(top / vh),
    right: clamp01(r / vw),
    bottom: clamp01(bottom / vh),
  };
}

function clamp01(v: number): number {
  return Math.min(1, Math.max(0, v));
}
